### PR TITLE
Feature/admob 2021 stability in future usage

### PR DESCRIPTION
**Pull request by DellaBitta**
Fixes stability issues regarding race conditions in the AdMob SDK:
 - Ensure that future objects are fully created before launching async operations which use their handles.
 - Remove the blocking destructor of BannerView on Android. 
   - There was a race condition where the the future might be destroyed before the destroy method returns from Android causing an assertion in the Future implementation.
 - Remove the blocking destructors of Interstitial and Rewarded Ads on iOS.
   - These implementations don't invoke AdMob objects and so don't need to be run on the main thread.
 
***

FTL testing.  Local Testing. [Integration CI](https://github.com/firebase/firebase-cpp-sdk/actions/runs/1516389819)
***

### Type of Change
Place an `x` the applicable box:
- [X] Bug fix. Add the issue # below if applicable.
- [ ] New feature. A non-breaking change which adds functionality.
- [ ] Other, such as a build process or documentation change.
***

### CODE REVIEW

File: admob/integration_test/src/integration_test.cc [hunk 1]
```diff
@@ -795,6 +795,8 @@ TEST_F(FirebaseAdMobTest, TestBannerView) {
   // Set the listener.
   TestBoundingBoxListener bounding_box_listener;
   banner->SetBoundingBoxListener(&bounding_box_listener);
+  PauseForVisualInspectionAndCallbacks();
```
Comment (DellaBitta) [inline admob/integration_test/src/integration_test.cc:798]: Wait to ensure that any callback attempts have time to complete before we start testing the listener.

File: admob/src/android/banner_view_internal_android.cc [hunk 2]
```diff
@@ -390,21 +386,29 @@ Future<void> BannerViewInternalAndroid::Resume() {
 
 Future<void> BannerViewInternalAndroid::Destroy() {
   firebase::MutexLock lock(mutex_);
-  return InvokeNullary(kBannerViewFnDestroy, banner_view_helper::kDestroy);
+  FutureCallbackData<void>* callback_data =
```
Comment (DellaBitta) [inline admob/src/android/banner_view_internal_android.cc:389]: No longer use futures to track the destruction as they may be destroyed before the're completed. This had caused a crash on some Android runs.

TBH, the Applicatoin _should already have invoked_ `Destroy()` prior to this object's destruction.  Invoking it again here isn't required.  I guess it was put here to protect our users from accidentally leaking memory. 

There doesn't seem to be a good reason to block the destructor though.  Instead I'll start an asynchronous attempt to destroy the Java resources, if they haven't been destroyed already, and then return immediately here.

File: admob/src/common/admob_common.cc [hunk 3]
```diff
@@ -227,7 +227,9 @@ const char* GetRequestAgentString() {
 // Mark a future as complete.
 void CompleteFuture(int error, const char* error_msg,
                     SafeFutureHandle<void> handle, FutureData* future_data) {
-  future_data->future_impl.Complete(handle, error, error_msg);
+  if (future_data->future_impl.ValidFuture(handle)) {
+    future_data->future_impl.Complete(handle, error, error_msg);
```
Comment (DellaBitta) [inline admob/src/common/admob_common.cc:231]: Just an extra safety check.

File: admob/src/ios/interstitial_ad_internal_ios.mm [hunk 4]
```diff
@@ -32,29 +32,20 @@
 
 InterstitialAdInternalIOS::~InterstitialAdInternalIOS() {
   firebase::MutexLock lock(mutex_);
-  // Clean up any resources created in InterstitialAdInternalIOS.
```
Comment (DellaBitta) [inline admob/src/ios/interstitial_ad_internal_ios.mm:35]: This code was waiting to run on the main thread but all it was doing was setting values to null.  While all AdMob operations should be conducted on the main thread, we're not actually performing any AdMob operations here, so let's simplify this code and just assign the null values here.

File: admob/src/ios/rewarded_ad_internal_ios.mm [hunk 5]
```diff
@@ -33,22 +33,13 @@
 RewardedAdInternalIOS::~RewardedAdInternalIOS() {
   firebase::MutexLock lock(mutex_);
   // Clean up any resources created in RewardedAdInternalIOS.
-  Mutex mutex(Mutex::kModeNonRecursive);
```
Comment (DellaBitta) [inline admob/src/ios/rewarded_ad_internal_ios.mm:36]: Simplify in the same way that we did for Interstitials, above.

File: admob/src_java/com/google/firebase/admob/internal/cpp/BannerViewHelper.java [hunk 6]
```diff
@@ -142,41 +142,57 @@ public void initialize(Activity activity) {
...
-        mPopUp = null;
+              synchronized (mPopUpLock) {
+                if (mPopUp != null) {
+                  mPopUp.dismiss();
+                  mPopUp = null;
+                }
+              }
+              synchronized (mBannerViewLock) {
+                if (destructor_invocation == false) {
+                  notifyBoundingBoxChanged(mBannerViewInternalPtr);
```
Comment (DellaBitta) [inline admob/src_java/com/google/firebase/admob/internal/cpp/BannerViewHelper.java:173]: Don't invoke the bounding box listener if this destroy() invocation originated from the C++ object destructor.  The bounding box listener won't be there anymore.

File: admob/src/android/banner_view_internal_android.cc (hunk 2 above)
Comment (jonsimantov) [reply admob/src/android/banner_view_internal_android.cc:389]: Yeah, I think there's no reason to actually wait for the Java destruction to complete.

**Review by jonsimantov — APPROVED** [review]

**Pull request merged**